src/bio/bio_recovery.c

```diff
@@ -0,0 +1,250 @@
+/**
+ * (C) Copyright 2018-2019 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B620873.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+#define D_LOGFAC	DD_FAC(bio)
+
+#include <spdk/io_channel.h>
+#include <spdk/blob.h>
+#include "bio_internal.h"
+
+/*
+ * The BIO blobstore (mapped to one NVMe device) owner xstream polls the
+ * state periodically, and takes predefined reaction routines on state
+ * transition, all the reaction routines must be non-blocking, otherwise
+ * the DAOS progress ULT will be blocked, and NVMe device qpair won't be
+ * polled.
+ */
+struct bio_reaction_ops	*ract_ops;
+
+/*
+ * Return value:	0: Faulty reaction is done;
+ *			1: Faulty reaction is in progress;
+ *			-ve: Error;
+ */
+static int
+on_faulty(struct bio_blobstore *bbs)
+{
+	int	tgt_ids[BIO_XS_CNT_MAX];
+	int	tgt_cnt, i, rc;
+
+	/* Transit to next state if faulty reaction isn't registered */
+	if (ract_ops == NULL || ract_ops->faulty_reaction == NULL)
+		return 0;
+
+	ABT_mutex_lock(bbs->bb_mutex);
+	tgt_cnt = bbs->bb_ref;
+	D_ASSERT(tgt_cnt <= BIO_XS_CNT_MAX && tgt_cnt > 0);
+
+	for (i = 0; i < tgt_cnt; i++)
+		tgt_ids[i] = bbs->bb_xs_ctxts[i]->bxc_xs_id;
```

Comment:
maybe bxc_xs_id can be renamed to bxc_tgt_id

Reply:
Yes, I'm going to do that in a separate cleanup patch later. (rename it in both BIO and SMD)